### PR TITLE
[java] Metrics framework: first metrics + first rule 

### PR DESCRIPTION
**Pull request by oowekyala**
Followup on the metrics framework.

### New metrics: NCSS, LOC, CYCLO, WMC
* Non commenting source statements (`NcssMetric`): differs from the version currently used by PMD by fully complying with <a href="http://www.kclee.de/clemens/java/javancss/">JavaNCSS rules</a>. The current implementation counts variable declarations such as `int x, y = 2, z;` as 3 statements while JavaNCSS counts only one statement. This can be amended later, and is subject to discussion 
* Lines Of Code (`LocMetric`)
* Cyclomatic complexity (`CycloMetric`): implemented following what was discussed in #445
* Weighed Method Count (`WmcMetric`)

### New rule: `CyclomaticComplexityRule`
This may replace what is atm 3 rules (see #445). It's also a PoC and sample use case for MetricVersion and ResultOption (described later on), and the framework as a whole.

### Temporary ruleset
I created a ruleset (`metrics.xml`) to test the behaviour of the rules the framework produces. The rules it contains can be merged back when the framework is mature (next major version). It currently contains only the new `CyclomaticComplexityRule`. 

#### Create and use variants of a metric: `MetricVersion`
* All metrics have a `STANDARD` version defined in `Metric.Version`, which is used as default.
* Metrics define additional variants in an enum nested inside their implementation class and are responsible for the correct handling of the `MetricVersions` that they define (see [`CycloMetric`](https://github.com/oowekyala/pmd/blob/metrics/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/metrics/CycloMetric.java#L71) for an example). 
* The user can compute a variant of a metric using an overload of `Metrics.get` (see [`CyclomaticComplexityRule`](https://github.com/oowekyala/pmd/blob/metrics/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java#L83) for an example).
* Different versions of the same metric can be cached at the same time thanks to `ParameterizedMetricKey`, which is now used to index memoization maps (see ClassStats & OperationStats)

#### Compute statistics about the value of a metric in a class: `ResultOption`
For example one can get the highest value of cyclomatic complexity among the operations of a class in one line: `Metrics.get(OperationMetricKey.CYCLO, classNode, ResultOption.HIGHEST)`. 
* All operation metrics benefit from that with no changes in their API
* The logic for the calculation is located in [`ClassStats`](https://github.com/oowekyala/pmd/blob/metrics/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/ClassStats.java#L205), and uses memoized results to speed up the computation.
* The results of those computations are not memoized but could be (though I'm not sure it's useful)

### Miscellaneous
* Improved getter and setter detection (see [`OperationSignature`](https://github.com/oowekyala/pmd/blob/metrics/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/signature/OperationSignature.java#L105)). That code is vastly experimental (understand 'hacky'), and I only tested it on positive examples. I'll add tests with negative examples in the next PR
*  The capability handling of metrics is wrapped around the metrics' computation: the `Metric` interface has a `supports(AccessNode)` method for which `AbstractMetric` provides a default implementation
* As usual some files have moved around:
  * AbstractMetric is out of `oom.metrics` and into `oom`, so that it can provide a hook on `Metrics.getTopLevelPackageStats()`, which is package-private, to its subclasses
  * Interfaces and the keys were moved into `oom.api` to declutter `oom`.

## Next steps
* I'd like to use the rule testing framework to test metrics individually. I think I could make one rule per metric, which would only print the value of the metric on each node, and a ruleset for those rules, then use the ordinary test framework. I could put that all into `src/test` so that it doesn't clutter the main codebase. The only thing is that I'll have to adapt the framework so that it looks for the ruleset somewhere in `src/test`.
**Edit:** another difficult thing is actually to get the ruleset parser to look for the classes of the rules in `src/test`. I don't think we can adapt that...

* I need to test the memoization infrastructure, until now I've just assumed it works
* I'll probably implement some more of the metrics that can be computed using AST exploration (eg TCC). They're quite easy to write, and the more metrics we have, the more rules we can write, and the more bugs we'll catch

That should take me the week. After that I'll move on to more complicated metrics (ATFD for one) and try to put signature matching to use. Then I can refactor GodClassRule.

Cheers!

### CODE REVIEW

**Comment by jsotuyod** [comment]
@adangel I assume this will make it into 5.8.0. If so, remember to add proper info on the changelog.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/metrics/xml/CyclomaticComplexity.xml [hunk 1]
```diff
@@ -0,0 +1,289 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<test-data>
+<!--
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/metrics/xml/CyclomaticComplexity.xml]: Many test cases are still commented out. Are these work-in-progress or can we already enable them?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+            if (classCyclo >= reportLevel || classHighest >= reportLevel) {
+                addViolation(data, node,
+                             new String[] {"class", node.getImage(), classCyclo + " (Highest = " + classHighest + ')'});
+            }
+        }
+        return data;
+    }
+
+    // TODO:cf consider merging these two methods (changes to JavaParserVisitor)
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java]: Hum ok. Maybe a simpler approach would be to use an adapter that defines `visit(ASTMethodOrConstructorDeclaration, Object)` and delegates to the concrete nodes.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/metrics/xml/CyclomaticComplexity.xml (hunk 1 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/metrics/xml/CyclomaticComplexity.xml]: That was a lapse, I commented them to ease debugging but they should work. I'll correct that

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/signature/OperationSignature.java [hunk 3]
```diff
@@ -67,19 +84,115 @@ public int hashCode() {
...
-                || node.getName().startsWith("set"))) {
-                return GETTER_OR_SETTER; // TODO:cf better getter or setter detection
+            } else if (isGetterOrSetter(node)) {
+                return GETTER_OR_SETTER;
             } else {
                 return METHOD;
             }
         }
+
+        private static boolean isGetterOrSetter(ASTMethodDeclaration node) {
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/signature/OperationSignature.java]: Right that might have gotten out of control. I thought looking at the code might yield better recall, but I think precision must be impacted, and the rules are difficult to read. Your approach seems more systematic and less prone to false positives.

I'll refactor that in the next PR

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        if (showMethodsComplexity && cyclo >= reportLevel) {
+            addViolation(data, node, new String[] {"method", node.getQualifiedName().getOperation(), "" + cyclo});
+        }
+        return data;
+    }
+
+
+    @Override
+    public Object visit(ASTConstructorDeclaration node, Object data) {
+        int cyclo = (int) Metrics.get(OperationMetricKey.CYCLO, node);
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java]: Right!

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java (hunk 2 above)
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java]: That will be for the next PR

**Comment by oowekyala** [comment]
@adangel If you agree, I'll work on your last comments for the next PR. By the way I managed to create a test ruleset for metrics and some testing rules, your advice on the classpath made it work!

**Comment by adangel** [comment]
@oowekyala Yes, that's fine. Glad you got the testing working!
